pyspinw/basis.py

```diff
@@ -0,0 +1,75 @@
+""" Functions for working with different basis vectors"""
+
+import numpy as np
+from pyspinw.dimensionality import dimensionality_check
+
+@dimensionality_check(vectors=(-1, 3))
+def find_aligned_basis(vectors: np.ndarray, rcond: float | None = None) -> tuple[np.ndarray, np.ndarray, np.ndarray]:
+    """ Find a set of orthonormal basis vectors aligned with the first being aligned to the input vectors
+
+    :param vectors: an n-by-3 numpy array containing the vectors to use to calculate the basis
+    :param rcond: tolerance on the calculation, default is 3 * machine precision
+    """
+
+    if rcond is None:
+        rcond = 3*np.finfo(vectors.dtype).eps # N * floating point epsilon
+
+    # Lengths, for checking for zeros and for normalising
+    lengths = np.sqrt(np.sum(vectors**2, axis=1))
+
+    #
+    # First basis vector will is just a normalised version of the input
+    #
+
+    # Avoid zero divisions by using np.divide instead of /
+    zero_vectors = lengths < rcond
+    e_1 = np.divide(vectors, lengths.reshape(-1, 1), where=~zero_vectors.reshape(-1, 1))
+
+    # Assign zero vectors to z
+    e_1[zero_vectors, 0] = 0.0
+    e_1[zero_vectors, 1] = 0.0
+    e_1[zero_vectors, 2] = 1.0
```

Comment:
I think that `e_1[zero_vectors,:] = np.array([[0., 0., 1.]])` will work here. Also in lines 43-45.

Reply:
done